### PR TITLE
Optimise integer parsing

### PR DESCRIPTION
**Pull request by fishcakez**
* Use single binary match context
* Do minimal pattern matches and maths

### CODE REVIEW

File: lib/elixir/lib/integer.ex [hunk 1]
```diff
@@ -238,67 +238,51 @@ defmodule Integer do
...
-  defp do_parse(<<char, rest::binary>>, base) do
-    if valid_digit_in_base?(char, base) do
-      do_parse(rest, base, parse_digit(char))
-    else
-      :error
+  digits = [{?0..?9, -?0}, {?A..?Z, 10-?A}, {?a..?z, 10-?a}]
+
+  for {chars, diff} <- digits, char <- chars do
+    defp parse_digits(<<unquote(char), rest::binary>>, base, sign)
+         when base > unquote(char+diff) do
```
Comment (whatyouhide) [inline lib/elixir/lib/integer.ex]: Spaces around `+` as well :bowtie:

File: lib/elixir/lib/integer.ex [hunk 2]
```diff
@@ -238,67 +238,51 @@ defmodule Integer do
   def parse("", base) when base in 2..36,
     do: :error
 
-  def parse(binary, base) when is_binary(binary) and base in 2..36 do
-    parse_in_base(binary, base)
+  def parse(<<bin::binary>>, base) when base in 2..36 do
+    parse_sign(bin, base)
```
Comment (whatyouhide) [inline lib/elixir/lib/integer.ex]: Would

```elixir
{sign, rest} = parse_sign(bin)
parse_digits(rest, base, sign)
```

work in the same way? We can keep it a bit more tidy by not calling `parse_digits/3` from `parse_sign/2`. But maybe we do bad stuff with the binary context this way 🤔

File: lib/elixir/lib/integer.ex [hunk 3]
```diff
@@ -238,67 +238,51 @@ defmodule Integer do
...
+  defp parse_sign(<<rest::binary>>, base) do
+    parse_digits(rest, base, :+)
   end
 
-  defp do_parse(<<char, rest::binary>>, base) do
-    if valid_digit_in_base?(char, base) do
-      do_parse(rest, base, parse_digit(char))
-    else
-      :error
+  digits = [{?0..?9, -?0}, {?A..?Z, 10-?A}, {?a..?z, 10-?a}]
```
Comment (lexmag) [inline lib/elixir/lib/integer.ex]: We need spaces around `-` and `+`. :bowtie:

File: lib/elixir/lib/integer.ex (hunk 2 above)
Comment (fishcakez) [inline lib/elixir/lib/integer.ex]: That would lose a binary match context optimisation.

Comment (whatyouhide) [inline lib/elixir/lib/integer.ex]: I feared so. Sad! It makes the code a bit convolute and it's really hard to get why just by reading it. Maybe we can add a comment about why it's done this way, not sure. We could in the meantime rename `parse_sign/2` to `parse_sign_and_digits/2`, wdyt?

**Comment by fishcakez** [comment]
I have fixed the spaces.

File: lib/elixir/lib/integer.ex (hunk 2 above)
Comment (fishcakez) [inline lib/elixir/lib/integer.ex]: Actually we can keep the single context here. Also ill move the parse_sign/2 logic into parse/2.

**Comment by fishcakez** [comment]
Simplifed the logic by removing the sign argument.

Simple benchmark run as `elixir -r file.exs`:

```elixir
defmodule Bench do

  def bench(n) do
    {time, _} = :timer.tc(__MODULE__, :test, [n])
    div(time, n)
  end

  def test(n) do
    binary = :binary.copy(<<"123456790">>, 100)
    
    _ =for _ <- 1..n do
      Integer.parse(binary)
    end
    
    :ok
  end
end

IO.inspect Bench.bench(1000)
```
x21 faster on OTP 19.2.1 : 4662 -> 222.

File: lib/elixir/lib/integer.ex (hunk 3 above)
Comment (lessless) [inline lib/elixir/lib/integer.ex]: Is there any justification to use variable on the module level? Obviously it's used in the metaprogramming block below, but shouldn't it be a module attribute at least?

**Pull request merged**